### ISSUE TITLE
`friendly_tangents` can show misleading view for structured matrix wrappers

### ISSUE THREAD

**Issue by sunxd3** [chalk-lab/Mooncake.jl#937]
@willtebbutt brought up this potential issue.

When using `friendly_tangents=true`, the `tangent_to_primal!!` conversion wraps the tangent's underlying matrix data directly into a `Symmetric` wrapper. Since the internal tangent data is not symmetric (only one triangle holds the gradient), the conversion produces misleading output where the gradient appears in both triangles.

```julia
using LinearAlgebra, Mooncake

f(S::Symmetric) = S[1, 2]
S = Symmetric([1.0 2.0; 2.0 4.0])
rule = Mooncake.build_rrule(f, S)
```

```julia
julia> _, grad_raw = Mooncake.value_and_pullback!!(rule, 1.0, f, S)
(2.0, (Mooncake.NoTangent(), Mooncake.Tangent{@NamedTuple{data::Matrix{Float64}, uplo::Mooncake.NoTangent}}((data = [0.0 1.0; 0.0 0.0], uplo = Mooncake.NoTangent()))))

julia> _, grad_friendly = Mooncake.value_and_pullback!!(rule, 1.0, f, S; friendly_tangents=true)
(2.0, (f, [0.0 1.0; 1.0 0.0]))

julia> grad_friendly[2]
2×2 Symmetric{Float64, Matrix{Float64}}:
 0.0  1.0
 1.0  0.0
```

## Problem

The raw tangent correctly stores `1.0` only at `[1,2]` (the stored triangle). When `tangent_to_primal!!` wraps this in `Symmetric`, the `[2,1]` entry is mirrored from `[1,2]`, making the gradient appear as `1.0` in both positions.

This is misleading: the user sees a symmetric gradient `[0 1; 1 0]` when the actual gradient information is only in the upper triangle `[0 1; 0 0]`.

## Round-trip behavior

The round-trip itself is correct because `primal_to_tangent!!` reads from the underlying storage, not the `Symmetric` view:

```julia
using LinearAlgebra, Mooncake

S = Symmetric([1.0 2.0; 2.0 4.0])
t_original = Mooncake.zero_tangent(S)
t_original.fields.data[1,2] = 1.0
```

```julia
julia> p = Mooncake.tangent_to_primal!!(deepcopy(S), t_original)
2×2 Symmetric{Float64, Matrix{Float64}}:
 0.0  1.0
 1.0  0.0

julia> p.data
2×2 Matrix{Float64}:
 0.0  1.0
 0.0  0.0

julia> t_roundtrip = Mooncake.zero_tangent(S)
Mooncake.Tangent{@NamedTuple{data::Matrix{Float64}, uplo::Mooncake.NoTangent}}((data = [0.0 0.0; 0.0 0.0], uplo = Mooncake.NoTangent()))

julia> Mooncake.primal_to_tangent!!(t_roundtrip, p)
Mooncake.Tangent{@NamedTuple{data::Matrix{Float64}, uplo::Mooncake.NoTangent}}((data = [0.0 1.0; 0.0 0.0], uplo = Mooncake.NoTangent()))
```

But this can still be confusing.

The same issue applies to `Hermitian` and maybe other structured matrix wrappers.

**Comment by Technici4n** [comment]
This is because of the `show` implementation used by the Symmetric matrix, right?

I don't know what we can really do about this. Maybe we should add a docs page explaining derivatives wrt. Symmetric types. I think the Enzyme devs worked on some related docs recently, and I'm sure it has similar behavior.

**Comment by yebai** [comment]
The issue is that, in these cases, `user_friendly_tangent_type(tangent_type(P)) != P`. While the primal type `P` coincides with `user_friendly_tangent_type` in most situations, these exceptions suggest a gap in the current design. 

One possible remedy would be to introduce an interface `user_friendly_tangent_type(tangent_type(P))` (or `user_friendly_tangent_type(P)` for convenience), which by default returns `P`, allowing users to override or customize the user-friendly tangent type via overloading.

**Comment by yebai** [comment]
I commented in https://github.com/chalk-lab/Mooncake.jl/pull/972#discussion_r2722428435: 

> `Mooncake.tangent_to_primal!!(primal, tangent)` may be worth revisiting in the future—for example, replacing it with an interface such as `Mooncake.friendly_tangent!!(friendly_tangent, tangent)`. In hindsight, using primal types to represent tangents is not entirely natural.

One possible approach would be to define `Mooncake.friendly_tangent!!` with a generic fallback to `tangent_to_primal!!`, and then allow users to override `friendly_tangent!!` for specific types: 

```julia
Mooncake.friendly_tangent!!(primal, tangent) = Mooncake.tangent_to_primal!!(primal, tangent)
Mooncake.friendly_tangent!!(::Symmetric, tangent)  # return a dense array instead
Mooncake.friendly_tangent!!(primal, t::NoTangent) = t  # return `NoTangent()`
```

These are schematic examples and do not capture all details (e.g., recursive behaviour), but I think this direction is worth discussing. Comments welcome!

cc @lkdvos too. 

EDIT 1: simplified suggestion.

**Comment by Technici4n** [comment]
I am not necessarily convinced. Enzyme should have the same behavior when it comes to symmetric matrices, and it's the only one that is consistent with finite differences, as unfortunate as it is

**Comment by lkdvos** [comment]
I have to say that I'm even a bit confused as to why this is an issue.

If I forget for a second about how these objects are layed out in memory, and think of their abstract meaning only, I feel like everything seems correct:
If we start from a `Symmetric` input to some function, why would the gradient not be symmetric? I.e. I would expect that if I have a loss function `f(x::Symmetric) -> Real`, I would be able to optimize over symmetric matrices by doing some naive gradient descent like:
```julia
for iter in 1:maxiter
    x -= alpha * gradient(f, x) # take a small step but remains symmetric
end
```
In other words, if I actually take this tangent and now add it to a dense array, it should indeed add to entries above and below the diagonal?
In fact, I would even have expected that if I would redefine the loss function to first convert to a dense array, the reverse pass of that conversion should exactly be a projection onto the symmetric subspace.

I understand that it is slightly misleading that the entries are stored in the upper/lower triangle and printed everywhere, so the memory layout is a bit weird, but this is already the case for the primal type as well, so I don't even think this is all that surprising behavior?

I might really be missing the point here though, so feel free to correct me!

**Comment by Technici4n** [comment]
@lkdvos the gradient of `f(A::Symmetric) = sum(A)` will for example be `[1 2; 1 0]` which then gets displayed as `[1 2; 2 1]`, instead of the expected `[1 1; 1 1]`.

**Comment by lkdvos** [comment]
Ah thanks, that example does actually make sense.

I guess then this is actually more problematic than simply a misleading `show`, since then my previous example of the naive gradient descent with friendly tangents would simply be wrong, because naively calling `+` on this generated friendly tangent will increment as `[1 2; 2 1]`.

Actually, investigating this a bit further, it seems like even `Mooncake._add_to_primal` doesn't correctly handle this:

```julia
using Mooncake
using DifferentiationInterface: DifferentiationInterface as DI

backend = DI.AutoMooncake(; config = Mooncake.Config(; friendly_tangents = false))
f(x) = x[2, 1]
val, grad = DI.value_and_gradient(f, backend, Symmetric([1.0 2.0; 2.0 1.0]))

julia> Mooncake._add_to_primal(Symmetric(zeros(2, 2)), grad)
2×2 Symmetric{Float64, Matrix{Float64}}:
 0.0  1.0
 1.0  0.0
```

Interestingly though, this is only when working with these types at the interface level, if you push it one level down the problem is gone:
```julia
g(x) = f(Symmetric(x))
val, grad = DI.value_and_gradient(g, backend, [1.0 2.0; 2.0 1.0])

julia> Mooncake._add_to_primal(zeros(2, 2), grad)
2×2 Matrix{Float64}:
 0.0  1.0
 0.0  0.0
```

---

Thinking a bit about solutions here, I don't think I would be a huge fan of having different types for the friendly tangent system here.
I feel like the conversion to primal should actually just be "fixed" for these structured matrices, in the sense that I would expect that I would expect the gradient of `f` to be `[0.0 0.5; 0.5 0.0]`, i.e. scaling the off-diagonal entries by a factor `1/2`.
Would that solve the issue?

I'm definitely still somewhat confused about the entire matter, it *feels* like something somewhere is off or inconsistent that we would have to do this to begin with, but I can't really tell what

**Comment by Technici4n** [comment]
Be careful about what Julia is printing. In your examples I would recommend reading what is in the `data` field to see exactly what is happening. Mooncake is computing the correct derivative from a "structural" pov, in the sense that doing anything else at the level of `f` would make the composition `g(x) = f(Symmetric(x))` give incorrect results. The problem is that you might indeed expect something else. Related: https://github.com/EnzymeAD/Enzyme.jl/issues/1334

**Comment by yebai** [comment]
`Symmetric`, `Hermitian`, and `SymTridiagonal` store a single value that represents
  multiple logical matrix positions — e.g. `S.data[i,j]` is read as both `S[i,j]` and                                                                                                 
  `S[j,i]`. The chain rule accumulates the full gradient for both reads into the one stored                                                                                             
  slot, which is internally correct. However, the `friendly_tangents=true` path should return
  a plain `Matrix{T}` rather than the internal structured type, and future work will apply                                                                                              
  the correct per-entry correction (halving off-diagonal entries to satisfy `⟨G, δS⟩ = δf`).
                                                                                                                                                                                        
  ## Fix                                                                     
                                                                                                                                                                                        
  Add two new extensible functions and update the `friendly_tangents` path to use them:

  - **`build_friendly_tangent(x)`** — allocates the output buffer for the friendly gradient.                                                                                            
    Returns `nothing` by default (primal type is already the friendly type). Override to
    return a pre-allocated buffer of the appropriate type — e.g. `Matrix{T}` for                                                                                                        
    `Symmetric{T}`.                                                          
  - **`tangent_to_friendly!!((primal, dest), tangent)`** — top-level conversion. The
    `(primal, dest)` bundle pairs dispatch metadata with the output buffer: `dest = nothing`                                                                                            
    falls back to `tangent_to_primal_internal!!`; a typed `dest` (e.g. `Matrix{T}`) enables
    in-place writes without allocating. A 2-argument convenience form                                                                                                                   
    `tangent_to_friendly!!(primal, tangent)` is also provided, which calls   
    `build_friendly_tangent(primal)` to obtain `dest`.                                                                                                                                  
                                                                             
  For `Symmetric{T,S}`, `Hermitian{T,S}`, `SymTridiagonal{T,V}`:                                                                                                                        
                                                                             
  - `build_friendly_tangent` returns a pre-allocated `Matrix{T}`                                                                                                                        
  - `tangent_to_friendly!!` copies the raw backing storage into that buffer in-place
                                                                                                                                                                                        
  In `interface.jl`, the `Cache` struct gains `tangent_as_primals` and `tangent_as_friendly`
  fields. At run time, `(primal, dest)` pairs are reconstructed and passed element-wise to                                                                                              
  `tangent_to_friendly!!` via `tuple_map`, replacing the previous `tangent_to_primal!!` call                                                                                            
  sites on the `friendly_tangents=true` path. The interim `tangent_to_primal_internal!!`
  specialisations on https://github.com/chalk-lab/Mooncake.jl/pull/1102 are dropped.
